[#6] feat(order): 주문 서비스 엔티티 및 리포지토리 추가 & 테스트 코드 작성

atto08 opened the pull request:
## #️⃣연관된 이슈

> - close: #6 

## 📝작업 내용

> 주문 엔티티, 리포지토리 추가 & 엔티티 테스트 코드 작성

- 엔티티
  - 주문
  - 주문 상품
- 리포지토리
  - 주문
- 테스트 코드
  - 주문 생성
  - 주문 상품 생성
  - 주문 상품목록 지정
  - 주문 상태 변경
 
### 

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
- 아키텍처 폴더 구조랑 코드 전체적으로 한번 피드백 부탁드립니다! 
- 테스트 코드 작성이 익숙치않아 이 방향이 맞는지 아니면 의견이나 방향성 주시면 적극 수용하겠습니다!!
- 엔티티에 주석처리 한 친구들은 어떤 타입으로 작성 하실지 아직 명확하지 않아서 주석처리 해놨습니다.
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?

atto08 committed:
feat(order): 의존성 추가

atto08 committed:
feat(order): 주문 상태 Enum 추가

atto08 committed:
feat(order): Order, OrderItem 엔티티 생성

atto08 committed:
feat(order): OrderCoreApplication 설정

atto08 committed:
feat(order): Order 리포지토리 생성

atto08 committed:
feat(order): Order 리포지토리 구현체 생성

atto08 committed:
feat(order): Order Dto 생성(request, response)

atto08 committed:
feat(order): Order 생성, 조회 테스트 코드 추가

yunjae62 commented on common/src/main/java/radiata/common/domain/order/dto/request/OrderCreateRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+public class OrderCreateRequestDto {
+
+    private String address;
+    private String comment;
+    private List<OrderItemCreateRequestDto> itemList;
+}
```
`class`가 아닌 `record` 타입으로 구현해주세요!!
`record` 타입은 자바 언어 차원에서 getter, equals, hashcode, toString, 불변을 보장해주는 문법입니다!!

atto08 replied:
레코드 타입 들어 만 봤는데 이번기회에 제대로 사용해보겟씁니다!

wonowonow commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+import radiata.service.order.core.domain.repository.OrderRepository;
+
+@SpringBootTest
+class OrderTest {
```
**통합 테스트의 경우 model Test 가 아닌, Service 등 비즈니스 디렉토리에 위치하는 게 좋아보입니다.**

우선 해당 테스트 코드는 엔티티 테스트 코드로 보이지 않고, Ksuid 나 JPA 의 동작을 확인하는 테스트 코드로 보입니다.
현도님이 작성해주신 Order 와 OrderItem 을 토대로 model Test 을 진행하면 Entity 내의 메서드가 정확히 동작하는지 다른 의존성과 관련없이 테스트 할 수 있을 거 같아요. `create`, `setOrderItems`, `updateOrderStatus` 에 관한 테스트 코드를 작성해주시는 건 어떨까요?

추가로 JPA 관련 테스트나 Ksuid 관련 테스트를 진행 하실 수는 있긴하나, 그건 저희가 담당하는 게 아닌 해당 라이브러리 개발자가 해야 좋을 거 같아요. 저희는 저희가 짠 코드가 저희 예상대로 (Exception, Success Cases)정상적으로 돌아가는지 확인하면 될 거 같습니다!

atto08 replied:
제가 작성한 건 통합테스트 였군여..
말씀해주신 피드백 천천히 읽어보니까 조금 이해된거 같습니다!
원호님이랑 재윤님 코드 참고해서 수정해보겠습니다!!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/model/entity/OrderItem.java:
```diff
@@ -0,0 +1,65 @@
+import lombok.NoArgsConstructor;
+
+@Entity
+@Table(name = "p_order_item")
```
테이블 접두사가 `p_`가 아니라 `r_` 입니다!!

atto08 replied:
앗 바로 수정햇씁니다 ! ㅎㅎ

wonowonow commented on common/src/main/java/radiata/common/domain/order/dto/request/OrderCreateRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+@AllArgsConstructor
+public class OrderCreateRequestDto {
```
Dto 의 같은 경우 `Record Class` 를 쓰는 것이 좋아보입니다.
Dto 는 값이 변경 될 일이 없어서 불변 객체를 쉽게 작성하도록 도와주는 Record 클래스를 이용하시는 게 어떨까요?
Record 에 익숙하지 않으시다면 private final 을 쓰셔도 무방합니다!

atto08 replied:
불변 객체를 쉽게 작성하게 도와주는 Record 클래스 메모햇습니다! 
이번기회에 마스터 하겠습니다 :)

wonowonow commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/model/entity/OrderItem.java:
```diff
@@ -0,0 +1,65 @@
+import lombok.NoArgsConstructor;
+
+@Entity
+@Table(name = "p_order_item")
```
r_order_item 으로 바꿔주세요!

atto08 replied:
정말 부끄러운 실수네요 ㅎㅎ..

yunjae62 commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+import org.springframework.boot.test.context.SpringBootTest;
+import radiata.service.order.core.domain.repository.OrderRepository;
+
+@SpringBootTest
+class OrderTest {
+
+    @Autowired
+    private OrderRepository orderRepository;  // 생성자 대신 필드 주입
+
```
패키지를 보니 Order 엔티티의 테스트 같은데요, OrderRepository는 여기서는 관심사가 아니라서 유닛 테스트만 진행하시면 됩니다!!
`@SpringBootTest`를 붙이면 실제 테스트 실행 시 스프링이 실행이 됩니다. 그래서 테스트 속도가 느려지게 돼요. 그리고 Order 엔티티의 기능들만 테스트하는 것이 관심사라서 `@Autowired private OrderRepository orderRepository` 역시 없어도 될 것 같습니다!!

atto08 replied:
스프링이 실행되는 것도 처음알았습니다! 피드백 적용하겠습니다!

yunjae62 commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+
+    private Order order;
+
+    @BeforeEach
+    @DisplayName("Order Entity 생성 테스트")
```
`@BeforeEach`에 `@DisplayName("Order Entity 생성 테스트")`를 붙이면 매 `@Test` 붙은 메서드마다 이 메소드가 실행되게 되고, 아마 그 개수만큼 `@DisplayName("Order Entity 생성 테스트")`가 뜰 것 같아서 제외하는 게 좋을 것 같은데요, 확실치 않아서 한번 확인해보시는 게 좋을 것 같습니다!
그리고 테스트 파일에서 cmd + N 누르시면 set up 으로 만드시면 beforeEach 자동으로 만들어줍니다!

atto08 replied:
꿀팁 감사합니다!! cmd + N 메모 햇씁니다 ㅎㅎ

yunjae62 commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+        System.out.println("---------------------------------------------");
+        System.out.println("#########[ 주문한 상품 리스트 정보]#########");
+        for (OrderItem item : order.getItemList()) {
+            System.out.println("Order Item ID: " + item.getId());
+            System.out.println("Order ID: " + item.getOrder().getId());
+            System.out.println("Product ID: " + item.getProductId());
+            System.out.println("Coupon Id " + item.getCouponIssuedId());
+            System.out.println("Reward Id " + item.getRewardPointId());
+            System.out.println("Quantity: " + item.getQuantity());
+            System.out.println("Unit Price: " + item.getUnitPrice());
+            System.out.println("Payment Price: " + item.getPaymentPrice());
+            System.out.println();
```
`System.out.println()`으로 테스트 하는 것이 아닌, `junit`과 `assertj` 라이브러리로 자동으로 테스트를 진행합니다!!
저랑 원호님이 작성한 테스트 보시면 (`Assertions`가 `import static`으로 `assertj`로 되어있고) `assertThat(...).isEqualsTo(...)`로 되어 있는 것을 볼 수 있습니다! 이거 참고하시면 됩니다!

atto08 replied:
이해를 잘 못해서 편의대로 해봤는데 오답이였네요!..
재윤님 원호님 코드 참고해서 수정해볼게요!!

yunjae62 reviewed (commented):
고생하셨습니다!!

그리고 오른쪽 위에 Reviewers에 저희 팀원분들 체크해두면 메일로 PR 메일 알림와서 바로 확인할 수 있습니다!! 체크 안 하면 알림이 안 와요.. 

또 여기에 수정하고 다 수정해서 재리뷰 받고 싶으시면 오른쪽 위에 Reviewers의 각 팀원 이름 옆에 새로고침 버튼이 생길텐데요, 그거 누르시면 다시 메일 가서 리뷰받고 싶다는 걸 저희가 알 수 있습니다!!

wonowonow commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        // Assert 문으로 특정 값 확인
+        assertNotNull(order, "Order should not be null");
```
**Assertions 의 값 비교 메서드와 와 Given When Then 을 사용해 테스트 코드를 작성해주세요.**

제가 생각하는 Spring 에서의 좋은 테스트 코드는 Assertions 를 사용해서 값 비교를 통해 메서드가 내부 로직이 잘못된 쪽으로 변경 되어도 즉시 발견할 수 있어야 한다고 생각합니다. 하지만 위의 코드는 그렇지 않습니다. 메서드 로직이 이상해져도 위의 테스트 코드는 통과될 수도 있어요!

보기 좋은 테스트 코드는 저도 아직 어렵지만, Given When Then 을 썼을 때 조금은 더 보기 좋은 테스트 코드를 작성 할 수 있었습니다.
한 번 제 코드 봐주세요!

atto08 replied:
Given-when-then 피드백 토대로 수정해보겠습니다!!

wonowonow commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/model/entity/OrderTest.java:
```diff
@@ -0,0 +1,104 @@
+
+        // Order 저장 (OrderItems는 cascade로 함께 저장)
+        orderRepository.save(order);
+    }
```
**Class 단위에 Transaction 이나, Before Each 를 사용해서 각 테스트 환경이 독립적으로 실행 될 수 있게 해주세요**

각 테스트는 독립적인 환경에서 변수없이 돌아가는 것이 중요합니다.
독립적인 환경이 아니면 다른 테스트 코드 추가 시 기존에 잘 돌아가던 로직이 동작하지 않을 수 있습니다.

atto08 replied:
설명해주시니까 원호님 테스트 코드에 구분지어 놓으신 의미가 보이네요!!
피드백 참고해서 수정해보겟습니다!

wonowonow reviewed (commented):
고생하셨습니다. 제가 코멘트 남긴 부분 확인해주세요!

wonowonow commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/model/vo/OrderStatus.java:
```diff
@@ -0,0 +1,22 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum OrderStatus {
```
OrderStatus 는 위치는 Value Object 디렉토리보다는 Constant 디렉토리를 따로 만드시는 게 좋아보입니다!

yunjae62 replied:
넵!

atto08 replied:
오호 vo 가 뭔지 잘 모르고 사용한거 같습니다 
수정 완료햇습니다!

hgalchi reviewed (commented):
assertj으로 test code를 작성하면 CI workflows도 통과될 듯 합니다! 고생많으셨습니다:)

atto08 committed:
refactor(order): 테이블 명 'p_order_item'으로 변경 - 리뷰 반영

atto08 committed:
refactor(order): DTO 클래스 record 타입으로 변경 - 리뷰 반영

atto08 committed:
refactor(order): OrderStatus 디렉토리 위치 이동(constant) - 리뷰 반영

atto08 committed:
refactor(order): Order 엔티티 테스트 코드 수정 - 리뷰 반영
1. SpringBootTest 제거 + 도메인(엔티티) 관심사만 적용
2. 통합테스트 x 👉단위 테스트 진행
3. JUnit & assertj 라이브러리를 통한 자동 테스트

atto08 committed:
refactor(order): 엔티티 생성 메서드 이름 변경 create -> of

atto08 committed:
refactor(order): 테스트 코드 given - when -then 수정 & 사용하지 않는 assertions 제거

atto08 committed:
[#14] refactor(order): order 엔티티 객체 생성 orderPrice 초기 값 0으로 지정 & setOrderPrice 추가

atto08 committed:
[#18] feat(order): 주문 생성 responseDto itemList 추가

atto08 committed:
[#18] feat(order): 서비스 메서드 생성
- 주문 아이디 생성
- 주문 생성
- 주문 상세 조회
- 주문 상태 변경

atto08 committed:
[#18] feat(order): 서비스 메서드 테스트 추가
- 주문 생성
- 주문 상태 변경

yunjae62 commented on common/src/main/java/radiata/common/domain/order/dto/response/OrderResponseDto.java:
```diff
@@ -0,0 +1 @@
    List<OrderItemResponseDto> itemList
) {

}
```
앗 이거 저만 이상하게 보이나요..??

atto08 replied:
오잉 저도 이상하게 나오네요 코드에선 문제없는데.. 
이부분도 수정해보겟습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/OrderCoreApplication.java:
```diff
@@ -0,0 +1,11 @@
+package radiata.service.order.core;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class OrderCoreApplication {
+    public static void main(String[] args) {
+        SpringApplication.run(OrderCoreApplication.class, args);
+    }
+}
```
core 부분에는 스프링 애플리케이션을 안 넣어야 합니다!! 
api 부분에 만드셔야합니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/service/OrderServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl {
```
특별히 구현하고 있는 인터페이스가 없다면 -Impl 없어도 될 것 같습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/service/OrderServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+    public Order getOrder(String orderId) {
+
+        return orderRepository.findById(orderId)
+            .orElseThrow(() -> new IllegalArgumentException("Order not found"));
```
`IllegalArgumentException` 에외를 쓰기보다, `BusinessException` 으로 `ExceptionMessage` 내부에 직접 예외를 만들어서 넣는 게 좋을 것 같습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/service/OrderServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+    public void updateOrderStatus(String orderId, OrderStatus updateStatus) {
+
+        Order order = orderRepository.findById(orderId)
+            .orElseThrow(() -> new IllegalArgumentException("Order not found"));
```
여기도 위에서 만든 예외를 적용하면 될 것 같습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/service/OrderServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+
+    // 주문 상세 조회
+    @Transactional(readOnly = true)
+    public Order getOrder(String orderId) {
+
+        return orderRepository.findById(orderId)
+            .orElseThrow(() -> new IllegalArgumentException("Order not found"));
+    }
```
그리고 이런 주문을 조회하는 구체적인 로직은 Implementation 계층에 넣으면 될 것 같습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/domain/service/OrderServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+
+    // 주문 생성
+    @Transactional
+    public Order createOrder(OrderCreateRequestDto createRequestDto, String orderId, String userId) {
+
+        Order order = Order.of(
+            orderId,
+            userId,
+            createRequestDto.address(),
+            createRequestDto.comment()
+        );
+
+        orderRepository.save(order);
+
+        return order;
+    }
```
실제 주문을 생성하는 구체적인 로직은 Implementation 계층으로 빼면 좋을 것 같습니다!

yunjae62 reviewed (commented):
수고하셨습니다!!!

atto08 committed:
[#14] refactor(order): order-core application 파일 제거

atto08 committed:
[#14] feat(order): 엔티티와 Dto간 변환을 위해 OrderMapper 추가

hgalchi commented on service/order/order-core/src/test/java/radiata/service/order/core/domain/service/OrderServiceImplTest.java:
```diff
@@ -0,0 +1,111 @@
+            // when: 주문 생성 로직 실행
+            Order createdOrder = orderService.createOrder(orderCreateDto, "order-01", "user-01");
+
+            // then: 주문 객체의 필드 값 확인
```
junit보단 assertj 사용을 추천드립니다! 
제공하는 인터페이스가 더 풍부하고  가독성 측면에서도 좋아 junit5에서도 assertj를 권장한다고 합니다!!

hgalchi reviewed (approved):
고생하셨습니다~- ̗̀( ˶' ３'˶) ̖́-

atto08 committed:
[#18] refactor(order): implementation 구현 및 service 구분

atto08 committed:
[#18] feat(order): 컴포넌트 스캔 설정 객체 추가

atto08 committed:
[#18] feat(order): 주문 implementation 테스트 -> service 테스트로 변경

atto08 committed:
[#18] refactor(order): Dto 공백 제거

atto08 committed:
[#18] refactor(order): OrderServiceTest Ci 실패 - db연결 문제 추정

atto08 committed:
[#18] refactor(order): OrderServiceTest Ci 실패 - db연결 문제 추정2

atto08 committed:
Merge branch 'main' into feat/order-core

atto08 committed:
[#18] refactor(order): OrderResponseDto 공백 제거 - mac 스타일(CR)

atto08 committed:
[#18] refactor(order): OrderResponseDto 공백 제거 - mac 스타일(LF)

atto08 committed:
[#18] refactor(order): 테스트 코드 라이브러리 변경
junit -> assertj

atto08 committed:
[#18] refactor(order): 테스트 코드 용 DB H2 설정

atto08 committed:
[#18] refactor(order): 테스트 코드 수정 - CI 시도

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/implemetation/OrderUpdater.java:
```diff
@@ -0,0 +1,18 @@
+import radiata.service.order.core.domain.model.constant.OrderStatus;
+import radiata.service.order.core.domain.model.entity.Order;
+
+@Implementation
+public class OrderUpdater {
+
+    // TODO 결제(대기중, 완료), 배송(중, 대기중, 완료) - 사용 예정
+    // 주문 상태 변경
+    @Transactional
+    public void updateOrderStatus( OrderStatus updateStatus, Order order) {
+
+        order.updateOrderStatus(updateStatus);
+    }
+}
```
이 부분은 Order에서 바로 처리할 수 있는 책임이라 Implementation 까지 안 가고 바로 서비스에서 order.updateOrderStatus(updateStatus); 하면 될 것 같습니다!

yunjae62 commented on service/order/order-core/src/main/java/radiata/service/order/core/implemetation/OrderUpdater.java:
```diff
@@ -0,0 +1,18 @@
+    // TODO 결제(대기중, 완료), 배송(중, 대기중, 완료) - 사용 예정
+    // 주문 상태 변경
+    @Transactional
+    public void updateOrderStatus( OrderStatus updateStatus, Order order) {
```
매개변수 부분에 띄어쓰기 하나가 들어가 있습니다!
설정-도구-저장 시 액션 -위에 3개 체크 하시면 자동으로 정렬이 됩니다!

yunjae62 reviewed (approved):
크리티컬한 문제는 없어 보입니다!

atto08 committed:
[#18] refactor(order): orderUpdateStatus implementation 제거
